### PR TITLE
bpo-40956: Convert _sqlite3.Cursor to Argument Clinic

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue40956

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Current `_sqlite3` AC conversion status:
- [x] Module level functions, except `connect()` (#22484)
- [x] Connection, except `__init__` (#23341 and #23838)
- [x] Row (#23964)
- [x] Cursor (this PR)
- [ ] Cache
- [ ] `_sqlite3.connect()` and `_sqlite3.Connection.__init__`

**Comment by erlend-aasland** [comment]
@corona10 / @serhiy-storchaka, would one of you like to review this? I've gone through the argument specs twice, but please take an extra look in case I've missed something.

File: Modules/_sqlite/cursor.c [hunk 1]
```diff
@@ -610,30 +609,61 @@ _pysqlite_query_execute(pysqlite_Cursor* self, int multiple, PyObject* args)
     }
 }
 
-PyObject* pysqlite_cursor_execute(pysqlite_Cursor* self, PyObject* args)
+/*[clinic input]
+_sqlite3.Cursor.execute as pysqlite_cursor_execute
+
+    sql: unicode
+    parameters: object = NULL
```
Comment (serhiy-storchaka) [inline Modules/_sqlite/cursor.c]:
```suggestion
    parameters: object(c_default='NULL') = ()
```

File: Modules/_sqlite/cursor.c [hunk 2]
```diff
@@ -610,30 +609,61 @@ _pysqlite_query_execute(pysqlite_Cursor* self, int multiple, PyObject* args)
...
-    return _pysqlite_query_execute(self, 0, args);
+    return _pysqlite_query_execute(self, 0, sql, parameters);
 }
 
-PyObject* pysqlite_cursor_executemany(pysqlite_Cursor* self, PyObject* args)
+/*[clinic input]
+_sqlite3.Cursor.executemany as pysqlite_cursor_executemany
+
+    sql: unicode
+    parameters: object = NULL
```
Comment (serhiy-storchaka) [inline Modules/_sqlite/cursor.c]: Actually the parameter is `seq_of_parameters`, and it is not optional.
```suggestion
    seq_of_parameters: object
```

File: Modules/_sqlite/cursor.c [hunk 3]
```diff
@@ -839,13 +888,49 @@ PyObject* pysqlite_cursor_fetchall(pysqlite_Cursor* self, PyObject* args)
     }
 }
 
-PyObject* pysqlite_noop(pysqlite_Connection* self, PyObject* args)
+/*[clinic input]
+_sqlite3.Cursor.setinputsizes as pysqlite_cursor_setinputsizes
+
+    sizes: object = None
```
Comment (serhiy-storchaka) [inline Modules/_sqlite/cursor.c]: Is it optional?

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: Apparently, but I find it weird. BTW, neither `setinputsizes` nor `setoutputsize` are documented.
```
$ python3.9
Python 3.9.1 (v3.9.1:1e5d33e9b9, Dec  7 2020, 12:10:52) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> con = sqlite3.connect(":memory:")
>>> cur = con.cursor()
>>> cur.setinputsizes()
>>> 
```

**Comment by erlend-aasland** [comment]
FYI, Ubuntu tests fail at `test_ttk_guionly`, not at `test_sqlite`.

**Comment by erlend-aasland** [comment]
The macOS check at https://github.com/python/cpython/runs/1625929931 triggered by e9535b83a13acb1b7b9b63ba43cbb57834d2e9cb should've failed: the sqlite3 module was not built! (The fix, commit 64b9c36fa0072b39310c6cbaa2142f9448bbea58, was pushed right after e9535b83a13acb1b7b9b63ba43cbb57834d2e9cb)

I guess this should be reported as a CI bug, @serhiy-storchaka?

File: Modules/_sqlite/cursor.c [hunk 4]
```diff
@@ -631,7 +631,7 @@ pysqlite_cursor_execute_impl(pysqlite_Cursor *self, PyObject *sql,
 _sqlite3.Cursor.executemany as pysqlite_cursor_executemany
 
     sql: unicode
-    parameters: object = NULL
+    seq_of_parameters as parameters: object
```
Comment (serhiy-storchaka) [inline Modules/_sqlite/cursor.c]: ` as parameters` is not needed. Just rename the parameter.

File: Modules/_sqlite/cursor.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/_sqlite/cursor.c]: They are documented in the Python DB API Spec ([PEP 249](https://www.python.org/dev/peps/pep-0249/) if there are no newer documents).

File: Modules/_sqlite/cursor.c (hunk 4 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: No, you'll need to change line 635 as well. You can't just rename the parameter without changing the parameter users as well.

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: I'll of course change both lines if you prefer that to explicitly naming the parameter in AC.

File: Modules/_sqlite/cursor.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: > They are documented in the Python DB API Spec ([PEP 249](https://www.python.org/dev/peps/pep-0249/) if there are no newer documents).

That's true, but they're not documented here: https://docs.python.org/3/library/sqlite3.html

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: > Apparently, but I find it weird.

On the other side, passing `None` as the `sizes` parameter is allowed according to the DB API, so defaulting to `None` is perhaps not so weird after all.

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: … but the DB API actually specifies that it's not optional (no brackets around the parameters)

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: According to the DB API, we should remove the default argument, but if we want to be backwards compatible, we shouldn't. But the function is a no-op, so I guess "backwards compatibility" is a no-op as well :)

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: See 935140d

File: Modules/_sqlite/cursor.c (hunk 4 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: See ~feac1a121aee59797008e4264156bd8708d28b42~ 829f23a

**Comment by erlend-aasland** [comment]
PTAL, @serhiy-storchaka. Any idea why `test_ttk_guionly` on Ubuntu keeps failing? I can't see that that test uses sqlite3 in any way...

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks for reviewing, @serhiy-storchaka !

**Comment by serhiy-storchaka** [comment]
`test_ttk_guionly` is something unstable test. Sometimes (rarely) it fails on buildbots.

But currently `test_nntplib` is failed because of network issues.

**Comment by erlend-aasland** [comment]
FYI, I'm ready with the PR for `_sqlite3.Cache` when this gets merged, @serhiy-storchaka. It's a pretty small PR. The diff stat is `3 files changed, 43 insertions(+), 18 deletions(-)` excluding autogenerated clinic code.

**Comment by serhiy-storchaka** [comment]
Please merge with master to fix test failures.

**Comment by erlend-aasland** [comment]
> Please merge with master to fix test failures.

Done. Is there anything more you need from me before merging?

**Review by berkerpeksag — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, @berkerpeksag

**Pull request merged**